File: android/app/src/main/kotlin/com/evanjt/traintime/ui/onboarding/CalloutBubble.kt
package com.evanjt.traintime.ui.onboarding

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Button
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.evanjt.traintime.LocalAppPalette

// A small anchored coach-mark, not a full-screen card. Title and body only: the
// Back/Skip/Next controls live in a fixed bar (TourNavBar) so they never move.
// The caret hints at the spotlighted feature; the host places the bubble in the
// screen half opposite the target so it never covers what it explains.
@Composable
fun CalloutBubble(
    title: String,
    body: String,
    caretUp: Boolean,
    modifier: Modifier = Modifier,
) {
    val bubbleColor = MaterialTheme.colorScheme.surfaceContainerHigh
    val bubbleBorder = BorderStroke(1.dp, MaterialTheme.colorScheme.outline.copy(alpha = 0.4f))
    Column(
        modifier = modifier.widthIn(max = 380.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
    ) {
        if (caretUp) Caret(up = true, color = bubbleColor)

        Surface(
            shape = RoundedCornerShape(16.dp),
            color = bubbleColor,
            border = bubbleBorder,
            shadowElevation = 8.dp,
        ) {
            Column(Modifier.padding(horizontal = 18.dp, vertical = 16.dp)) {
                Text(
                    title,
                    color = MaterialTheme.colorScheme.onSurface,
                    fontSize = 17.sp,
                    fontWeight = FontWeight.SemiBold,
                )
                Spacer(Modifier.size(6.dp))
                Text(
                    body,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                    fontSize = 14.sp,
                )
            }
        }

        if (!caretUp) Caret(up = false, color = bubbleColor)
    }
}

// The tour's fixed navigation bar: progress dots + Back / Skip / Next, pinned to
// the bottom so the controls stay in the same place on every page. The Back slot
// keeps its width when hidden so Next never shifts.
@Composable
fun TourNavBar(
    index: Int,
    total: Int,
    onBack: () -> Unit,
    onSkip: () -> Unit,
    onNext: () -> Unit,
    nextLabel: String,
    modifier: Modifier = Modifier,
) {
    val palette = LocalAppPalette.current
    Surface(
        modifier = modifier.fillMaxWidth(),
        shape = RoundedCornerShape(16.dp),
        color = MaterialTheme.colorScheme.surfaceContainerHigh,
        border = BorderStroke(1.dp, MaterialTheme.colorScheme.outline.copy(alpha = 0.4f)),
        shadowElevation = 8.dp,
    ) {
        Row(
            Modifier.padding(horizontal = 10.dp, vertical = 6.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Box(Modifier.width(64.dp)) {
                if (index > 0) TextButton(onClick = onBack) { Text("Back") }
            }
            Row(
                horizontalArrangement = Arrangement.Center,
                verticalAlignment = Alignment.CenterVertically,
                modifier = Modifier
                    .weight(1f)
                    .clearAndSetSemantics { contentDescription = "Step ${index + 1} of $total" },
            ) {
                repeat(total) { i ->
                    val active = i == index
                    Box(
                        Modifier
                            .padding(horizontal = 3.dp)
                            .size(if (active) 8.dp else 6.dp)
                            .background(
                                if (active) palette.platform else MaterialTheme.colorScheme.onSurfaceVariant.copy(alpha = 0.4f),
                                CircleShape,
                            ),
                    )
                }
            }
            TextButton(onClick = onSkip) {
                Text("Skip", color = MaterialTheme.colorScheme.onSurfaceVariant)
            }
            Button(onClick = onNext) { Text(nextLabel) }
        }
    }
}

// A simple triangle caret drawn with a rotated square, pointing toward the target.
@Composable
private fun Caret(up: Boolean, color: Color) {
    androidx.compose.foundation.Canvas(
        modifier = Modifier
            .padding(vertical = 2.dp)
            .size(width = 18.dp, height = 9.dp),
    ) {
        val path = androidx.compose.ui.graphics.Path().apply {
            if (up) {
                moveTo(size.width / 2f, 0f)
                lineTo(size.width, size.height)
                lineTo(0f, size.height)
            } else {
                moveTo(0f, 0f)
                lineTo(size.width, 0f)
                lineTo(size.width / 2f, size.height)
            }
            close()
        }
        drawPath(path, color)
    }
}

File: apple/TrainTimePhone/Views/Onboarding/CalloutBubble.swift
import SwiftUI

// A small anchored coach-mark, not a full-screen card. The caret hints at the spotlighted
// feature; the host places the bubble in the screen half opposite the target so it never sits
// over what it explains. Peer of the Android CalloutBubble.kt.
struct CalloutBubble: View {
    let title: String
    let message: String
    let index: Int
    let total: Int
    let caretUp: Bool
    let nextLabel: String
    let onBack: () -> Void
    let onSkip: () -> Void
    let onNext: () -> Void

    var body: some View {
        VStack(spacing: 2) {
            if caretUp { Caret(up: true) }

            VStack(alignment: .leading, spacing: 6) {
                Text(title)
                    .font(.headline)
                Text(message)
                    .font(.subheadline)
                    .foregroundStyle(.secondary)

                HStack(spacing: 4) {
                    if index > 0 {
                        Button("Back", action: onBack)
                            .buttonStyle(.borderless)
                    }
                    HStack(spacing: 6) {
                        ForEach(0..<total, id: \.self) { i in
                            Circle()
                                .fill(i == index ? Color.accentColor : Color.secondary.opacity(0.4))
                                .frame(width: i == index ? 8 : 6, height: i == index ? 8 : 6)
                        }
                    }
                    .frame(maxWidth: .infinity)
                    .accessibilityElement()
                    .accessibilityLabel("Step \(index + 1) of \(total)")

                    Button("Skip", action: onSkip)
                        .buttonStyle(.borderless)
                        .tint(.secondary)
                    Button(nextLabel, action: onNext)
                        .buttonStyle(.borderedProminent)
                        .controlSize(.small)
                }
                .padding(.top, 8)
            }
            .padding(.horizontal, 18)
            .padding(.vertical, 16)
            .background(
                RoundedRectangle(cornerRadius: 16)
                    .fill(Color(uiColor: .tertiarySystemBackground))
                    .shadow(radius: 8, y: 2)
            )
            .overlay(
                // Faint border so the callout reads as a raised sheet over the dimmed background.
                RoundedRectangle(cornerRadius: 16)
                    .stroke(Color.primary.opacity(0.12), lineWidth: 1)
            )

            if !caretUp { Caret(up: false) }
        }
        .frame(maxWidth: 380)
    }
}

// A triangle caret pointing toward the target.
private struct Caret: View {
    let up: Bool
    var body: some View {
        Canvas { context, size in
            var path = Path()
            if up {
                path.move(to: CGPoint(x: size.width / 2, y: 0))
                path.addLine(to: CGPoint(x: size.width, y: size.height))
                path.addLine(to: CGPoint(x: 0, y: size.height))
            } else {
                path.move(to: CGPoint(x: 0, y: 0))
                path.addLine(to: CGPoint(x: size.width, y: 0))
                path.addLine(to: CGPoint(x: size.width / 2, y: size.height))
            }
            path.closeSubpath()
            context.fill(path, with: .color(Color(uiColor: .tertiarySystemBackground)))
        }
        .frame(width: 18, height: 9)
        .accessibilityHidden(true)
    }
}
